packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,280 @@
+        //// 3. Count the number of eligible proposals.
+        //// 4. Count the number of votes in eligible proposals.
+
+        uint256 numEligibleProposals;
+        for (uint256 i; i < proposals.length; ++i) {
+            // make sure proposal finished voting
+            uint endBlock = max(proposals[i].endBlock, proposals[i].objectionPeriodEndBlock);
+            require(block.number > endBlock, 'all proposals must be done with voting');
+
+            // skip non eligible proposals
+            if (proposals[i].forVotes < (proposals[i].totalSupply * proposalEligibilityQuorumBps_) / 10_000) {
+                delete proposals[i];
+                continue;
+            }
+
+            // proposal is eligible for reward
+            ++numEligibleProposals;
+
+            uint256 votesInProposal = proposals[i].forVotes + proposals[i].againstVotes + proposals[i].abstainVotes;
+            t.numEligibleVotes += votesInProposal;
+        }
+
+        //// Check that distribution is allowed:
+        //// 1. At least one eligible proposal.
+        //// 2. One of the two conditions must be true:
+        //// 2.a. Number of eligible proposals is at least `numProposalsEnoughForReward`.
+        //// 2.b. At least `minimumRewardPeriod` seconds have passed since the last update.
+
+        require(numEligibleProposals > 0, 'at least one eligible proposal');
+        if (numEligibleProposals < params.numProposalsEnoughForReward) {
+            require(
+                lastProposal.creationTimestamp > lastProposalRewardsUpdate + params.minimumRewardPeriod,
+                'not enough time passed'
+            );
+        }
+        lastProposalRewardsUpdate = lastProposal.creationTimestamp;
+
+        // Calculate the reward per proposal and per vote
+        t.rewardPerProposal = proposalRewardForPeriod / numEligibleProposals;
+        t.rewardPerVote = votingRewardForPeriod / t.numEligibleVotes;
+
+        //// Second loop over the proposals:
+        //// 1. Skip proposals that were deleted for non eligibility.
+        //// 2. Reward proposal's clientId.
+        //// 3. Reward the clientIds that faciliated voting.
+        //// 4. Make sure all voting clientIds were included.
+
+        for (uint256 i; i < proposals.length; ++i) {
+            // skip non eligible deleted proposals
+            if (proposals[i].endBlock == 0) continue;
+
+            uint32 clientId = proposals[i].clientId;
+            if (clientId != 0) {
+                clientBalances[clientId] += t.rewardPerProposal;
+            }
+
+            uint256 votesInProposal;
+            NounsDAOV3Types.ClientVoteData[] memory voteData = proposals[i].voteData;
+            uint256 votes;
+            for (uint256 j; j < votingClientIds.length; ++j) {
+                clientId = votingClientIds[j];
+                votes = voteData[j].votes;
+                if (clientId != 0) {
+                    clientBalances[clientId] += votes * t.rewardPerVote;
```

Comment:
tstore: here and in the proposal client ID reward above, might want to try using tstore/tload, doesn't feel like a must to me

Reply:
same comment as above